### ISSUE TITLE
Numerical Stability of the Jacobian

### ISSUE THREAD

**Issue by valentinwust** [Kingrimursel/GeometricAutoencoder#1]
I'm trying to adapt your geometric loss to a standard NB RNA-expression autoencoder (like dca etc.). Which more or less works (after some small changes to deal with 1D inputs, hopefully correctly), but computing the Jacobian is pretty instable. So I need to do some ad-hoc rescaling of outputs to keep it within the 1e-20 to 1e20 range; var(log(det(J**2))) should be independent of that, but this is hit-or-miss. And I guess this makes sense, if I have 3000 output dimensions I would expect the Jacobian to scale very strongly with an overall factor. Did you ever encounter problems like that when doing your experiments, and do you have any hints how to best stay within a range where the numerics work out?

**Comment by phnazari** [comment]
Hey Valentin,
thank you very much for reaching out. We are excited about your application of the geometric regularizer!
While we agree that evaluating the determinant of `J^tD JD` on a mini-batch might contain some small or big outliers, we are still not sure that we understand the problem completely. We furthermore want to point you to the now published ICML2023 paper, which describes our Geometric Autoencoder: https://openreview.net/forum?id=VAEZ4nrqAk

Could you maybe provide us with a couple of details about how you think the described problem affects your results? Are there runtime errors? What exactly is your current hack?

Some thoughts from our side: After log-scaling, the outliers' effect on the loss should not be too strong, especially if there are only a couple of them. We have nevertheless decided to filter them out in our implementation, which gave good results.
In our code:
https://github.com/phnazari/GeometricAutoencoder/blob/e33a95c2a61e0f6e7c824b4dcaf859fc02e52a12/src/criterions.py#L77
Did you implement this filter? If yes, how many points are being filtered out?
One could, however, also think about collapsing all of the outliers into the most extreme tolerable values. Here, one would have to ensure that the backward pass still accumulates gradients.

On the practical side, we never worked with counts directly, but rather with pre-processed scRNAseq datasets. The problem of small values can be avoided by adding a small epsilon, i.e. calculating `log(x + eps)`.

We hope this helps!

**Comment by valentinwust** [comment]
First of all thank you for the paper, that was helpful. From the paper it looks like you implement autoencoders that work on 50–100 PCA components with an mse loss?

I could have probably been more explicit about what I am doing. What I was trying was applying your loss to an autoencoder that works on the count data with high (3000–4000 or so) dimensionality, like [dca](https://www.nature.com/articles/s41467-018-07931-2). This general approach is what is mostly used for RNA data, since it allows optimizing the log-likelihood of the noisy counts under an appropriate noise model (usually negative Binomial), and the results are easily interpretable since they are basically the smoothed expression of cells. (Using a reduced dimensionality as the input is probably still useful, but the output usually has a high dimension.) So I have an autoencoder that basically takes the 3000 dimensional input to the PCA (transcript counts, normalized and log1p, with or without norm to std 1 etc.), decodes it to 3000 logits, and the original counts are reconstructed as `softmax(logits) * total # of counts`. It might make sense to get the Jacobian from latent to logits either before or after softmax, but regardless the dimension increase is from 16 or so to 3000.

I'm basically using your code for your loss term, I think with some slight changes since it is set up for 2D data. And it kind of works, but I think the high dimensionality of the output is causing some problems. I only get a useful loss if I do some ad-hoc scaling of the output of the model (only for this loss, global scalings shouldn't change the log variance), and even this doesn't work that well. Because of the high dimensionality, a factor of 10 is sufficient to change the Jacobian determinants from 1e-15 up to 1e15, so it's hard to consistently stay in a range where the numerics doesn't break down.

I don't know how relevant this is, I was just curious how much this would change an autoencoder like this, and then wondered whether you had already though about this problem.

**Comment by phnazari** [comment]
Thank you very much for clarifying, now I understand your application a bit better. We are indeed training the autoencoder with an MSE loss, but our datasets have up to ~250 PCA components.

Is it correct that you want to scale the outputs to a certain range, because you hope that in this range the calculation of the Jacobian matrix is more stable? I am not sure this conclusion is true, but let us assume that it is for now. If it is, I do not understand why you are worried about the scaling of the Jacobian determinant itself then (end of third paragraph). 

That being said, I have a few more thoughts. Though you might already know this, my first thought is that your choice of loss should not limit the applicability of the geometric regularizer (see Appendix E.4, "Training with other Loss Functions"). So in particular, cross entropy should be fine.

I would furthermore recommend applying the geometric loss to the logits before they are passed through the softmax. The main motivation here is that the softmax does not have any free parameters, so it will not be affected by the side loss. Rather, it just creates unnecessary overhead in computing the decoder's Jacobian, while contributing to numerical instability.

Next, we want to see where the sensitivity to scaling of the output comes from.
Calculating the Jacobian matrix is a linear operation, so scaling the outputs by a factor of x should scale the Jacobian by the same factor. When multiplying the Jacobian with itself, we can expect the product to scale by a factor of x\*x. Hence, the determinant of the product scales by a factor of (x\*x)^(latent_dimension).
This shows two things.
1) It explains why scaling the output by a factor of 10 changes the values of the Jacobian determinant by 30 orders of magnitude - as in your example - if latent space is sixteen-dimensional.
2) More importantly, it shows that the scaling property is **independant** of the output dimension.

I am also not so sure about the effect of the high output dimension. You argue that it introduces numerical instability into the calculation of the Jacobian determinant, but I am not so sure this effect is significant. What PyTorch does after all is use the chain-rule to compute derivative of all of the 3000 outputs w.r.t. all of the 16 inputs. Hence, the output-dimensionality per se has no effect on the stability of the algorithm. Rather, rounding errors can accumulate and be reinforced by large network depth.

Hence, I propose you to try the following:
1) Try training with smaller latent dimension. This might keep your outputs in a reasonable manitude.
2) If you are worried about numerical (in)stability, try decreasing the depth of the decoder.

Of course, both of these fixes decrease performance.

Hope this helps. If you clarify my questions from the beginning, I will try to think about the problem more! :)

**Comment by valentinwust** [comment]
Your analysis of the scaling is of course correct, thank you for the in-depth engagement with my problem. Maybe I should have though about this a bit longer, it indeed works without problems when I use a low latent dimension. I wanted to test what happens to the normal higher-dimensional latent embedding with your loss, but this is probably not really a useful application anyway.

> I would furthermore recommend applying the geometric loss to the logits before they are passed through the softmax. The main motivation here is that the softmax does not have any free parameters, so it will not be affected by the side loss. Rather, it just creates unnecessary overhead in computing the decoder's Jacobian, while contributing to numerical instability.

However, I don't agree that this choice is irrelevant, since it does affect the metric on the original space.

**Comment by phnazari** [comment]
Great, let us know if you have any new findings or further questions.

For clarification: I should have been more precise. I did not mean to say that the choice is irrelevant, however I would stick to my recommendation. While the metric on the original space is fixed (it is the Euclidean metric), it does affect the pullback metric; in one case one also uses the softmax for pulling back the Euclidean metric, in the other case one does not.

**Comment by sdamrich** [comment]
Thank you both for this discussion. I think making the loss work seamlessly with higher latent space is a relevant point. I have three thoughts:
1. Take the d-th root of the determinant before the log, where d is the latent space dimension. This might make everything slightly more robust against the latent dimension. But probably the issue is in the determinant computation anyway. 
2. PyTorch has a [logdet function](https://pytorch.org/docs/stable/generated/torch.logdet.html). This might be more stable. Perhaps we should add a tiny bit of the identity to make sure the matrix is nicely invertible. 
3. Maybe this is what PyTorch's logdet function does under the hood, but we could also try out bypassing the determinant entirely like [here](https://blogs.sas.com/content/iml/2012/10/31/compute-the-log-determinant-of-a-matrix.html) via PyTorch's [Cholesky decomposition](https://pytorch.org/docs/stable/generated/torch.linalg.cholesky.html#torch.linalg.cholesky).

**Comment by valentinwust** [comment]
I haven't tested this extensively, but looks like simply using logdet fixes my issue.

**Comment by phnazari** [comment]
Now using PyTorch's [logdet](https://pytorch.org/docs/stable/generated/torch.logdet.html) for increased numerical stability.
